docs/2.guide/3.going-further/12.custom-router.md

```diff
@@ -0,0 +1,144 @@
+::
+
+### Using pages:extends hook
+
+You can add, change or remove pages from the scanned routes with the `pages:extend` nuxt hook. For example, to prevent creating routes for any `.ts` files:
+
+```ts [nuxt.config.ts]
+export default defineNuxtConfig({
+  hooks: {
+    'pages:extend' (pages) {
+      // add a route
+      pages.push({
+        name: 'profile',
+        path: '/profile',
+        component: () => import('~/pages/profile.vue')
+      })
+
+      // remove routes
+      function removePagesMatching (pattern: RegExp, pages: NuxtPage[] = []) {
+        const pagesToRemove = []
+        for (const page of pages) {
+          if (pattern.test(page.file)) {
+            pagesToRemove.push(page)
+          } else {
+            removePagesMatching(pattern, page.children)
+          }
+        }
+        for (const page of pagesToRemove) {
+          pages.splice(pages.indexOf(page), 1)
+        }
+      }
+      removePagesMatching(/\.ts$/, pages)
+    }
+  }
+})
+```
+
+### Using a module
+
+If you plan to add a whole set of pages related with a specific functionality, you might want to use a [Nuxt module](/modules).
+
+The [Nuxt kit](docs/guide/going-further/kit) provides a few ways [to add routes](/docs/api/advanced/kit#pages):
+- extendPages (callback: pages => void)
+- extendRouteRules (route: string, rule: NitroRouteConfig, options: ExtendRouteRulesOptions)
+
+## Router Options
+
+It is possible to customize [vue-router options](https://router.vuejs.org/api/interfaces/routeroptions.html).
+
+### Using `app/router.options`
+
+It is possible to customize [vue-router options](https://router.vuejs.org/api/interfaces/routeroptions.html).
+
+This is the recommended way to specify router options.
+
+```js [app/router.options.ts]
+import type { RouterConfig } from '@nuxt/schema'
+
+// https://router.vuejs.org/api/interfaces/routeroptions.html
+export default <RouterConfig> {
+}
+```
+
+### Using `nuxt.config`
+
+**Note:** Only JSON serializable options are configurable:
+
+- `linkActiveClass`
+- `linkExactActiveClass`
+- `end`
+- `sensitive`
+- `strict`
+- `hashMode`
+
+```js [nuxt.config]
+export default defineNuxtConfig({
+  router: {
+    // https://router.vuejs.org/api/interfaces/routeroptions.html
+    options: {}
+  }
+})
+```
+
+### Hash Mode (SPA)
+
+You can enable hash history in SPA mode. In this mode, router uses a hash character (#) before the actual URL that is internally passed. When enabled, the **URL is never sent to the server** and **SSR is not supported**.
+
+```ts [nuxt.config.ts]
+export default defineNuxtConfig({
+  ssr: false,
+  router: {
+    options: {
+      hashMode: true
+    }
+  }
+})
+```
+
+#### Custom History (advanced)
+
+You can optionally override history mode using a function that accepts base url and returns history mode.
+If returning `null` or `undefined`, Nuxt will fallback to the default history.
```

Comment:
```suggestion
You can optionally override history mode using a function that accepts the base URL and returns history mode. If returning `null` or `undefined`, Nuxt will fallback to the default history.
```

Reply:
I may be more readable the way it is, it was the point of https://github.com/nuxt/nuxt/pull/19565/commits/30fd48f0338f59324ed4d61af93c924f86ca868d

Reply:
made a small edit - how about now?